raft/raft.hh

```diff
@@ -251,6 +251,19 @@ struct log_entry {
 
 using log_entry_ptr = seastar::lw_shared_ptr<const log_entry>;
 
+template <typename T>
+inline size_t memory_usage_of(const T& v) {
+    if constexpr(std::is_same_v<T, command>) {
+        return v.size() + sizeof(log_entry);
+    }
+    if constexpr(std::is_same_v<T, log_entry>) {
+        if (const auto* c = get_if<command>(&v.data); c != nullptr) {
+            return memory_usage_of(*c);
+        }
+    }
+    return 0;
```

Comment:
`wait_max_log_size` is now only called for commands, in `add_entry_on_leader`, I preserve this behaviour. Moreover, it takes `log_entry` as a parameter, which we don't have in `add_entry_on_leader`.

Reply:
I do not know why `set_configuration` does not check it. IIRC it did at some point. Probably `modify_config` refactoring broke it. We deliberately did not check log size for internally added non joint config to not block configuration change, but user initiated move to joint config should check the limit. This is for another patch though. 

I do not suggest to use entry_size() as is but it can be refactored to support both direct command or log entry.

Also I do not think memory usage should account for a log entry itself. When we drop a command this does not really release its log entry struct.

Reply:
> When we drop a command this does not really release its log entry struct.

True, but when we add command `log_entry` still consumes memory, if we don't account for it we could get OOM for single byte commands.

Reply:
Adding static log entry size to a one byte command does not improve the situation much and allows the log data structure to grow to a huge sized. The patch drops max entry size limit. I think it should remain.

Reply:
```
sizeof(raft::log_entry) == 136
default max_log_size_in_bytes = 4 * 1024 * 1024;
```
If we don't account for `log_entry`, memory consumption could reach 4 * 1024 * 1024 * 136 == 570mb for single byte commands, which could easily trigger OOM.

> The patch drops max entry size limit. I think it should remain.

You mean max log size. Keeping it adds complexity in code (another semaphor+code that handles it) and in config (another config element), we need a reason for it. What's the problem with huge in-memory logs, if the memory consumption is limited? It can take time to replay the log when node restarts, we could optimise it by storing `_applied_idx` in `store_commit_idx`.

Reply:
> ```
> sizeof(raft::log_entry) == 136
> default max_log_size_in_bytes = 4 * 1024 * 1024;
> ```
> 
> If we don't account for `log_entry`, memory consumption could reach 4 * 1024 * 1024 * 136 == 570mb for single byte commands, which could easily trigger OOM.

It is still just a constant factor. With the patch the problem is that after all those entries are freed the log still takes most of the memory.

> 
> > The patch drops max entry size limit. I think it should remain.
> 
>You mean max log size. Keeping it adds complexity in code (another semaphor+code that handles it) and in config (another config element), we need a reason for it.

The reason is that you do not what the log to grow to huge sized.

> What's the problem with huge in-memory logs, if the memory consumption is limited?

The memory is not reclaimable. You may have zero memory usage according to your internal accounting but in fact use almost "memory limit" bytes just by having an empty log.

Also my guess is the reason CI fails is because IIRC we have tests that relies on be able to specify log length in entries to create specific scenario to test. Not sure though.

> It can take time to replay the log when node restarts, we could optimise it by storing `_applied_idx` in `store_commit_idx`.

How is this related?

Reply:
> It is still just a constant factor. With the patch the problem is that after all those entries are freed the log still takes most of the memory.

It's a constant factor which makes memory accounting more accurate and helps avoid the OOM scenario above.

BTW, `boost::deque` [deallocates](https://github.com/boostorg/container/blob/develop/include/boost/container/deque.hpp#L2236) unused blocks, so memory is in fact reclaimable. The default `block_bytes == 512`, we should increase it probably.

Reply:
> How is this related?

We could load `_applied_idx` on `server_impl::start` and not have to reapply all the entries from the last snapshot.

Reply:
How is this related to this series? And no, this is not that simple if you do not want double apply unless you know how to save apply index atomically while applying entries.

Reply:
> > It is still just a constant factor. With the patch the problem is that after all those entries are freed the log still takes most of the memory.
> 
> It's a constant factor which makes memory accounting more accurate and helps avoid the OOM scenario above.

I am not interested in discussing contrived scenarios that try to prove that you can allocate more 1 bytes objects that 137 bytes objects. We are not operating on this kind of error margin here. There is no such thing as allocating 1 byte object anyway. It will always take much more. 

> 
> BTW, `boost::deque` [deallocates](https://github.com/boostorg/container/blob/develop/include/boost/container/deque.hpp#L2236) unused blocks, so memory is in fact reclaimable. The default `block_bytes == 512`, we should increase it probably.

Is this guaranty by the standard? Both boost and std deque has `shrink_to_fit`. If standard guaranties that all free memory is always freed I do not see what is the need for `shrink_to_fit`. OTOH may be we can call this `shrink_to_fit` on log truncation and free the memory explicitly. Although https://en.cppreference.com/w/cpp/container/deque/shrink_to_fit says "It is a non-binding request to reduce the memory usage without changing the size of the sequence. It depends on the implementation whether the request is fulfilled." which does not sound optimistic, but I guess it means that the memory usage will not be reduces beyond freeing free blocks. Non free blocks will not be shrunk.

Reply:
> How is this related to this series?

Before this patch the log was limited by the number of items, now this limit is in bytes, which means that now there can be many elements in the log and double apply can become a performance problem.

> And no, this is not that simple if you do not want double apply unless you know how to save apply index atomically while applying entries.

I meant not to get rid of the double apply completely, but to reduce it to the size of one batch in `applier_fiber`. Not sure about the profit here though, we can still get the entire log in one batch. On the other hand, `snapshot_threshold` imposes an approximate limit on the number of double applies, so it's probably fine.

Reply:
> I meant not to get rid of the double apply completely, but to reduce it to the size of one batch in `applier_fiber`. 

If you apply from a latest snapshot there is no double apply since a snapshot does not yet contain commands those commands.

Reply:
> I am not interested in discussing contrived scenarios that try to prove that you can allocate more 1 bytes objects that 137 bytes objects. We are not operating on this kind of error margin here. There is no such thing as allocating 1 byte object anyway. It will always take much more.

Ok, it's not one byte, it's [eight bytes](https://github.com/scylladb/seastar/blob/f9f5228b74f71c04892730c6721509ac722c122b/src/core/memory.cc#L1392), the smallest pool size there is still one byte though. As you said, this is a general purpose library, and it's perfectly fine to have eight byte commands.

> Is this guaranty by the standard?

I doubt it, but the boost deque implementation actually does it, here's a [relevant comment](https://github.com/boostorg/container/blob/develop/include/boost/container/deque.hpp#L1232).
Perhaps we should leave a comment next to `_log` field that this is another reason of using boost deque.

Reply:
> > I am not interested in discussing contrived scenarios that try to prove that you can allocate more 1 bytes objects that 137 bytes objects. We are not operating on this kind of error margin here. There is no such thing as allocating 1 byte object anyway. It will always take much more.
> 
> Ok, it's not one byte, it's [eight bytes](https://github.com/scylladb/seastar/blob/f9f5228b74f71c04892730c6721509ac722c122b/src/core/memory.cc#L1392), the smallest pool size there is still one byte though. As you said, this is a general purpose library, and it's perfectly fine to have eight byte commands.

You ignored first part of the sentence and bike shading the second one instead.

> 
> > Is this guaranty by the standard?
> 
> I doubt it, but the boost deque implementation actually does it, here's a [relevant comment](https://github.com/boostorg/container/blob/develop/include/boost/container/deque.hpp#L1232). Perhaps we should leave a comment next to `_log` field that this is another reason of using boost deque.

I am not interested in how one particular implementation handles it. If it is not guarantied by the standard it means it can change tomorrow. We should not rely on internals of a library. Also we use boost just because std one throws during move. If tomorrow we will find something better in abseil we may change to it. It is just a matter of changing one line of code. `shrink_to_fit` may provide a solution though.

Reply:
> You ignored first part of the sentence and bike shading the second one instead.

It's not bike shading, I just don't agree that this scenario is not important.

> I am not interested in how one particular implementation handles it. If it is not guarantied by the standard it means it can change tomorrow. We should not rely on internals of a library. Also we use boost just because std one throws during move.

I don't agree with this picture of the world. Any library has its advantages and disadvantages, they are documented and cannot change, simply because a lot of client scenarios will be broken. In the end, why use abseil, if not for the sake of its specific features?

Anyway, this discussion seems to be theoretical, do we have action items relevant to this pr?

Reply:
* extended the comment about boost::deque, that it releases memory when items are removed
* moved memory_usage_of to log.h
* not sure if we need to update `log_limiter_semaphor` on `set_configuration`, since there can only be one `set_configuration` in the fly. In any case it's a separate ticket/pr.